test/topology_custom/test_tablets.py

```diff
@@ -166,6 +166,35 @@ async def check_allocated_replica(expected: int):
         assert len(fragments[k]) == rf_to, f"Found mutations for {k} key on {fragments[k]} hosts, but expected only {rf_to} of them"
 
 
+# ALTER tablets KS cannot change RF of any DC by more than 1 at a time.
+# In a multi-dc environment, we can create replicas in a DC that didn't have replicas before,
+# but the above requirement should still be honoured, because we'd be changing RF from 0 to N in the new DC.
+# Reproduces https://github.com/scylladb/scylladb/issues/20039#issuecomment-2271365060
+# See also cql-pytest/test_tablets.py::test_alter_tablet_keyspace_rf for basic scenarios tested
+@pytest.mark.asyncio
+async def test_multidc_alter_tablets_rf(request: pytest.FixtureRequest, manager: ManagerClient) -> None:
+    config = {"endpoint_snitch": "GossipingPropertyFileSnitch", "enable_tablets": "true"}
+
+    logger.info("Creating a new cluster of 1 node in 1st DC and 2 nodes in 2nd DC")
+    await manager.server_add(config=config, property_file={'dc': f'dc1', 'rack': 'myrack'})
+    await manager.server_add(config=config, property_file={'dc': f'dc2', 'rack': 'myrack'})
+    await manager.server_add(config=config, property_file={'dc': f'dc2', 'rack': 'myrack'})
+
+    conn = manager.get_cql()
+    conn.execute("create keyspace ks with replication = {'class': 'NetworkTopologyStrategy', 'dc1': 1};")
+    conn.execute("create table ks.t (pk int primary key);")  # need a table to not only change schema, but also replicas
+    with pytest.raises(InvalidRequest, match="Cannot modify replication factor of any DC by more than 1 at a time."):
+        # changing RF of dc2 from 0 to 2 should fail
+        conn.execute("alter keyspace ks with replication = {'class': 'NetworkTopologyStrategy', 'dc2': 2};")
+
+    # changing RF of dc2 from 0 to 1 should succeed
+    conn.execute("alter keyspace ks with replication = {'class': 'NetworkTopologyStrategy', 'dc2': 1};")
+    # also check that we can safely remove all replicas by changing RF of dc2 from 1 to 0 back again
```

Comment:
Maybe go all the way to `2` first since we're already here?

Reply:
If you mean to increment the RF of `dc2` from 1 to 2, then I think that RF in-/decrementing for a single DC is already covered in `test_alter_tablet_keyspace_rf`.
What came to my mind, however, is that I never test in-/decrementing RF of 2 DCs at once - this should be still OK, if they the resulting RFs don't differ by more than 1. However, to my surprise, what I ended up discovering was something else, namely that if a DC is not explicitly listed in an ALTER, then this ommitted DC i zeroed, consider:
```
create keyspace ks with replication = {'class': 'NetworkTopologyStrategy', 'dc1': 1} // we only have dc1 : 1
alter keyspace ks with replication = {'class': 'NetworkTopologyStrategy', 'dc2': 1} // we only have dc2 : 1, so dc1 is gone
alter keyspace ks with replication = {'class': 'NetworkTopologyStrategy', 'dc1': 2, 'dc2': 2} // this fails, because dc1 would have RF changed from 0 to 2
```
I couldn't find any information about it in the docs or maybe I don't understand the docs. I found something relevant on `replication_factor` tag, but it's forbidden to be used when ALTERing tablets-enabled keyspaces. 
So now I wonder what's the correct behaviour - I assumed that ommitted DCs just won't have their RF changed.

Reply:
yes, the expected behavior is that the rf's specified in alter apply only to the modified dcs while omited dcs are unchanged.
You can test this with 5.4

Reply:
So I fixed the code that zeroed skipped DCs, and, to my surprise, again, incrementing RFs in 2+ DCs in a single ALTER crashes, consider:
```
INFO  2024-08-21 21:04:49,005 [shard 0:stmt] alter_keyspace - dc: dc1, new_rf: 2, old_rf: 1
INFO  2024-08-21 21:04:49,005 [shard 0:stmt] alter_keyspace - dc: dc2, new_rf: 2, old_rf: 1
INFO  2024-08-21 21:04:49,024 [shard 0: gms] migration_manager - Update Keyspace: KSMetaData{name=ks, strategyClass=org.apache.cassandra.locator.NetworkTopologyStrategy, strategyOptions={"dc1": "2", "dc2": "2"}, cfMetaData={}, durable_writes=true, tablets={"enabled":false}, userTypes=org.apache.cassandra.config.UTMetaData@0x511000413cf0}
...
ERROR 2024-08-21 21:04:49,033 [shard 0: gms] raft - [074b7785-e3d1-4b13-95b9-3e2b76803cde] applier fiber stopped because of the error: std::_Nested_exception<raft::state_machine_error> (State machine error at raft/server.cc:1340): std::_Nested_exception<std::runtime_error> (Failed to read tablet metadata): std::runtime_error (Too many pending replicas for table 3a682bf0-5ff0-11ef-88df-873840c8e5eb tablet 0: {260b7bb3-9775-40c9-8fd1-352634e32c4e:1, c1f0ba45-83e4-48f0-b946-8c3fb60e7c5c:0})
```
Since allowing more than 1 pending replicas is a bigger effort, I'm guessing now I should change the validation code in ALTER to check if RF of at most a single DC is not changed by more than 1? 
CC @bhalevy @tgrabiec

Reply:
Yup. Rebuilding a tablet may not have more than one pending replica

Reply:
OK, I am validating this precondition now (that we can change at most 1 RF by 1).